File: py/portfolio.py
import functools
from money import Money
from bank import Bank


class Portfolio:
    def __init__(self):
        self.moneys = []
        self._eur_to_usd = 1.2

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank, currency):
        sum = 0
        conversionFailures = []
        for m in self.moneys:
            try:
                sum += bank.convert(m, currency).amount
            except Exception as ex:
                conversionFailures.append(ex.args[0])

        if(len(conversionFailures) > 0):
            raise Exception("Missing exchange rate(s):[%s]" % (
                ",".join(conversionFailures)))

        return Money(sum, currency)

File: js/portfolio.js
const Money = require("./money");

class Portfolio {
    constructor() {
        this.moneys = [];
    }

    add(...moneys) {
        this.moneys = this.moneys.concat(moneys);
    }

    evaluate(currency) {
        let conversionFailures = [];
        let total = this.moneys.reduce((sum, money) => {
            const convertedAmount = this.convert(money, currency);
            if (convertedAmount === undefined) {
                conversionFailures.push(`${money.currency}->${currency}`);
                return sum;
            }
            return sum + convertedAmount;
        }, 0);

        if (conversionFailures.length > 0) {
            throw new Error(`Missing exchange rate(s):[${conversionFailures.join()}]`);
        }
        return new Money(total, currency);
    }

    convert(money, currency) {
        let exchangeRates = new Map();
        exchangeRates.set("EUR->USD", 1.2);
        exchangeRates.set("USD->KRW", 1100);

        if (money.currency === currency) {
            return money.amount;
        }

        const conversionKey = `${money.currency}->${currency}`;
        const rate = exchangeRates.get(conversionKey);
        if (rate === undefined) {
            return undefined;
        }
        return money.amount * rate;
    }
}

module.exports = Portfolio;